### PR TITLE
[diffusion] Refactor TeaCache

### PR DESCRIPTION
**Pull request by eitanturok**
## Motivation

This PR cleans up the TeaCache implementation.

1. Teacache Parameters were split between 4 different classes, updated at different times, and had a messy get_context function to glue it all together. Now we have 3 clearly seperated classes with well-defined API boundaries.
2. Adding Teacache to a new model now requires 5 lines instead of ~100.
3. Remove duplicated code for CFG pos + CFG neg
4. Implemented teacache more generically (no inheritance) so we can support different types of caching in the future
5. Remove if statements and numpy calls so more friendly for torch.compile

This is a prerequisite for supporting different timestep caching methods like MagCache (#18498 #19957).

## Modifications

Here are the changes in more detail:

### 1. **Fragmented State**

**The Problem:** Parameters were scattered across the model class, `TeaCacheMixin`, `TeaCacheParams`, and `TeaCacheContext`. The `TeaCacheParams` class confusingly had methods attached to it. These classes and parameters were all updated at different times, making it difficult to manage teacache across different requests.

**The Solution:** Split logic into three distinct components:
* **TeaCacheParams:** A pure data class for user settings (thresholds, offsets). It has no methods and is updated at the start of every new request by a user.
* **TeaCacheState:** Manages internal runtime data (cached tensors, accumulated $L1$ distances). The step counter is now moved here instead of being attached to the model. This is updated after every forward pass.
* **TeaCacheStrategy:** The actual implementation logic that takes in `TeaCacheParams` and `TeaCacheState` to decide when to skip a computation.

At a high level, this is what `TeaCacheStrategy` looks like. (In reality, the methods take in different arguments.)
```py
class TeaCacheStrategy:
	def __init__(TeaCacheParams):
	def maybe_reset(TeaCacheState, TeaCacheParams): # maybe reset the cache state
	def should_skip(TeaCacheState): # decide if we should should skip computing the forward pass for this timestep
	def read(TeaCacheState): # read from the cache
	def write(TeaCacheState): # write to the cache
```

### 2. **Integration Overhead**

**The Problem:** Adding TeaCache to a new model required ~100 lines of boilerplate code (e.g., the large amount of code deleted in `wanvideo.py`).

**The Solution:** Rewrote the TeaCache API so models now only need to pass in the modulated input. All other logic is generic and handled by `TeaCacheStrategy`.

### 3. **Code Duplication:**

**The Problem:** Handling positive and negative CFG branches required duplicated code blocks in teacache to track separate caching logic.

**The Solution:** We have a different `TeaCacheState` state for positive and negative CFG branches.

### 4. **Cannot support multiple types of caching:**

**The Problem:** `Cacheable-DiT` inherited directly from `TeaCacheMixin`. Since we only know which cache type we want during the forward pass not at __init__, the model couldn't easily support or switch between different caching strategies as we'd need to change the class we are inheriting from.

**The solution:** Instead of `Cacheable-DiT` inheriting the cache, just assign the cache to `Cacheable-DiT.cache` (composition). It is now initialized via `init_cache()`, allowing the model to dynamically support multiple caching types.

### 5. **Not torch.compile friendly:**

**The Problem:** TeaCache called `np.poly1d` and had if statements for deciding when to skip a forward pass. This would make it hard to compile this code.

**The Solution:** Use `torch.where` instead of `if` and implement `np.poly1d` in torch.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
10. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
11. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
12. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

### CODE REVIEW

File: python/sglang/multimodal_gen/configs/sample/sampling_params.py [hunk 1]
```diff
@@ -161,9 +161,8 @@ class SamplingParams:
 
     # TeaCache parameters
     enable_teacache: bool = False
-    teacache_params: Any = (
-        None  # TeaCacheParams or WanTeaCacheParams, set by model-specific subclass
-    )
+    cache_params: Any | None = None
+    calibrate_cache: bool = False
```
[the remark this answers is not in the document]
Comment (eitanturok) [reply python/sglang/multimodal_gen/configs/sample/sampling_params.py]: removed

File: python/sglang/multimodal_gen/configs/sample/sampling_params.py [hunk 2]
```diff
@@ -161,9 +161,8 @@ class SamplingParams:
 
     # TeaCache parameters
     enable_teacache: bool = False
-    teacache_params: Any = (
-        None  # TeaCacheParams or WanTeaCacheParams, set by model-specific subclass
-    )
+    cache_params: Any | None = None
```
[the remark this answers is not in the document]
Comment (eitanturok) [reply python/sglang/multimodal_gen/configs/sample/sampling_params.py]: undid

**Comment by eitanturok** [comment]
Can this get a review? Blocking magcache implementation...

File: python/sglang/multimodal_gen/configs/sample/teacache.py [hunk 3]
```diff
@@ -56,27 +56,3 @@ class TeaCacheParams(CacheParams):
         default=None, repr=False
     )
     use_ret_steps: bool | None = None
-
-    def get_coefficients(self) -> list[float]:
```
Comment (mickqian) [inline python/sglang/multimodal_gen/configs/sample/teacache.py:60]: hey, why do we remove this?

**Review by mickqian — COMMENTED** [review]
would you also mention this change in corresponding doc session?

File: python/sglang/multimodal_gen/configs/sample/teacache.py (hunk 3 above)
Comment (eitanturok) [reply python/sglang/multimodal_gen/configs/sample/teacache.py:60]: `get_coefficients()` was removed because `TeaCacheParams` is now a pure data class — it only holds user-set parameters, no logic. 

`TeaCacheStrategy` centralizes all the logic, so coefficient resolution lives in `TeaCacheStrategy.maybe_reset()` alongside the rest of the per-generation setup.

Comment (mickqian) [reply python/sglang/multimodal_gen/configs/sample/teacache.py:60]: I get the pure-data-class direction, but `get_coefficients()` still feels like config resolution rather than runtime logic. moving it into `TeaCacheStrategy.maybe_reset()` mixes parameter resolution with request lifecycle / state reset, so I don't think this is necessarily clearer.

**Review by mickqian — COMMENTED** [review]
A few behavior issues still look blocking before merge:

1. `TeaCacheStrategy.maybe_reset()` only increments `state.step` when `step == 0`, so the skip window and cross-request reset logic can drift.
2. `CachableDiT.init_cache()` only runs once when `self.cache is None`, so `enable_teacache` becomes sticky to the first request.
3. after splitting CFG into `state` / `state_neg`, `_get_skip_boundaries()` still multiplies the boundaries by 2, which no longer looks consistent with the new state model.
4. `weights_updater.py` checks `isinstance(module, TeaCacheStrategy)`, but `module` here is the model module, so cache flush will not trigger this.

that said, the refactor direction makes sense

File: python/sglang/multimodal_gen/configs/sample/teacache.py (hunk 3 above)
Comment (eitanturok) [reply python/sglang/multimodal_gen/configs/sample/teacache.py:60]: i moved `_get_coefficents` and `_get_skip_boundaries` back to `TeaCacheParams`

**Review by mickqian — COMMENTED** [review]
a small follow-up comment

File: python/sglang/multimodal_gen/runtime/cache/teacache.py [hunk 4]
```diff
@@ -1,316 +1,153 @@
...
-        coefficients: Polynomial coefficients for L1 rescaling.
-        teacache_params: Full TeaCacheParams for model-specific access.
+class TeaCacheState:
+    """Tracks step progress, cached tensors, and L1 distances for a single CFG path."""
+
+    step: int = 0
+    previous_modulated_input: torch.Tensor | None = None
+    previous_residual: torch.Tensor | None = None
+    accumulated_rel_l1_distance: torch.Tensor | None = None
+
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/cache/teacache.py:30]: this now assumes exactly 5 coefficients, but the doc example in `docs/diffusion/performance/cache/teacache.md` still shows 3. 
please either make the implementation handle variable-length polynomials again, or update the docs/example to match this contract, which I personally prefer the former

**Comment by eitanturok** [comment]
Addressed all your comments.

1. Step counter logic — update the step-counter every step at the start of  'should_skip() '. Deleted maybe_reset() for simplicity.

2. Cache initialization stickiness — fixed. 'maybe_init_cache()' creates a new cache or removes the cache at the start of every new generation (when 'get_forward_context().current_timestep' is 0.

3. CFG skip boundary *2 — removed. _get_skip_boundaries() now just resolves the raw num_inference_steps with no doubling. 

4. 'isinstance(module, TeaCacheStrategy)' in 'weights_updater.py' — replaced with 'hasattr(module, "cache") and module.cache is not None', so the cache flush works correctly with the composition pattern and handles the disabled-cache case.

5. '_rescale_distance_tensor' now handles variable length polynomials in a torch.compile friendly way.

Also took the opportunity to simplify TeaCacheStrategy while here: it now takes plain values at construction (no TeaCacheParams coupling), TeaCacheState is a dataclass, and maybe_reset() is removed since a fresh strategy object is created each generation anyway.

File: python/sglang/multimodal_gen/runtime/cache/teacache.py (hunk 4 above)
Comment (eitanturok) [reply python/sglang/multimodal_gen/runtime/cache/teacache.py:30]: '_rescale_distance_tensor' now handles variable length polynomials in a torch.compile friendly way.

File: python/sglang/multimodal_gen/runtime/loader/weights_updater.py [hunk 5]
```diff
@@ -213,8 +212,8 @@ def update_weights_from_disk(
 
         if success and flush_cache:
             for _, module in modules_to_update:
-                if isinstance(module, TeaCacheMixin):
-                    module.reset_teacache_state()
+                if hasattr(module, "cache") and module.cache is not None:
+                    module.cache.reset()
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/loader/weights_updater.py]: This looks like a real bug on the weight-update path: `TeaCacheStrategy` does not define `reset()`, so `module.cache.reset()` will raise here when `flush_cache=True` and TeaCache is enabled.

**Review by mickqian — COMMENTED** [review]
One follow-up inline comment.

File: python/sglang/multimodal_gen/runtime/loader/weights_updater.py (hunk 5 above)
Comment (eitanturok) [reply python/sglang/multimodal_gen/runtime/loader/weights_updater.py]: fixed

`weight_updater.py` now calls `cache.reset_states()` to create new `TeaCacheState` objects.

**Comment by eitanturok** [comment]
can you add run-ci @mickqian

**Comment by eitanturok** [comment]
In multimodal unit tests, I removed the 4 stale entries that used to test `do_cfg=True` behavior which doubled the number of steps since we removed `do_cfg` from `TeaCacheParams`.

**Comment by eitanturok** [comment]
I fixed a numerical stability issue in TeaCache polynomial rescaling.

The baseline tests were originally generated using np.poly1d (Python float64). However, my branch evaluated these polynomials in Pytorch using the model's default dtype which was bf16. This caused catastrophic cancellation since bf16 can't handle the very large terms in the polynomial evaluation. To fix this, I upcasted the polynomial evaluation to fp32.


The cmd
```bash
python -m pytest     python/sglang/multimodal_gen/test/server/test_server_1_gpu.py::TestDiffusionServerOneGpu::test_diffusion_generation[wan2_1_t2v_1.3b_teacache_enabled]
```
was failing on my branch but now it passes. My branch with casting now gets the same results as master whereas my branch without casting differed.

On `main`
```
[Consistency Check] wan2_1_t2v_1.3b_teacache_enabled: FAILED
  Thresholds: clip>=0.93, ssim>=0.92, psnr>=24.0, mean_abs_diff<=9.0
  Min similarity: 0.9193
  Min SSIM: 0.1901
  Min PSNR: 10.2002
  Max mean_abs_diff: 60.6102
  Frame details:
    Frame 0: clip=0.9406 ssim=0.1913 psnr=10.2498 mean_abs_diff=60.2920 FAIL
    Frame 1: clip=0.9193 ssim=0.1935 psnr=10.2177 mean_abs_diff=60.4762 FAIL
    Frame 2: clip=0.9346 ssim=0.1901 psnr=10.2002 mean_abs_diff=60.6102 FAIL
```

On my branch with casting [d51aab3](https://github.com/sgl-project/sglang/pull/21613/commits/d51aab3f40e196808133b9e8eac6b7d6645d3243)
```
[Consistency Check] wan2_1_t2v_1.3b_teacache_enabled: FAILED
  Thresholds: clip>=0.93, ssim>=0.92, psnr>=24.0, mean_abs_diff<=9.0
  Min similarity: 0.9193
  Min SSIM: 0.1901
  Min PSNR: 10.2002
  Max mean_abs_diff: 60.6102
  Frame details:
    Frame 0: clip=0.9406 ssim=0.1913 psnr=10.2498 mean_abs_diff=60.2920 FAIL
    Frame 1: clip=0.9193 ssim=0.1935 psnr=10.2177 mean_abs_diff=60.4762 FAIL
    Frame 2: clip=0.9346 ssim=0.1901 psnr=10.2002 mean_abs_diff=60.6102 FAIL
```
On my branch without casting [811233f](https://github.com/sgl-project/sglang/pull/21613/commits/811233fe124a9e9b8c3e20aee9c5251f40dcd3ef)
```
[Consistency Check] wan2_1_t2v_1.3b_teacache_enabled: FAILED
  Thresholds: clip>=0.93, ssim>=0.92, psnr>=24.0, mean_abs_diff<=9.0
  Min similarity: 0.9412
  Min SSIM: 0.2026
  Min PSNR: 10.4619
  Max mean_abs_diff: 58.6682
  Frame details:
    Frame 0: clip=0.9494 ssim=0.2053 psnr=10.4829 mean_abs_diff=58.4755 FAIL
    Frame 1: clip=0.9412 ssim=0.2090 psnr=10.4975 mean_abs_diff=58.3973 FAIL
    Frame 2: clip=0.9487 ssim=0.2026 psnr=10.4619 mean_abs_diff=58.6682 FAIL
```

**Comment by eitanturok** [comment]
I found another bug in the existing TeaCache implementation that was causing the wan2_1_t2v_1.3b_teacache_enabled test to fail. I fixed this in #24227.

Full test cmd
```PYTHONPATH=python python3 -m pytest -s -v --tb=short python/sglang/multimodal_gen/test/server/test_server_1_gpu.py -k wan2_1_t2v_1.3b_teacache_enabled```
